src/training/Trainer.cpp

```diff
@@ -101,54 +108,115 @@ struct Trainer {
+  {
+    or_json::mObject feature_param_tree;
+    std::stringstream ssparams;
+    ssparams << params.get < std::string > ("json_feature_params");
+
+    {
+      or_json::mValue value;
+      or_json::read(ssparams, value);
+      feature_param_tree = value.get_obj();
+    }
+
+    // Create the pointer depending on the type of descriptors
+    std::string feature_type = feature_param_tree["type"].get_str();
+
+#if CV_VERSION_MAJOR == 3
+    if (feature_type == "ORB")
+    {
+      int n_features = feature_param_tree["n_features"].get_uint64();
+      int n_levels = feature_param_tree["n_levels"].get_uint64();
+      float scale_factor = feature_param_tree["scale_factor"].get_real();
+
+      feature2d_ = cv::ORB::create(n_features, scale_factor, n_levels);
+    }
+    else if(feature_type == "AKAZE")
+    {
+      feature2d_ = cv::AKAZE::create();
+    }
+#else
+    if (feature_type == "ORB")
+    {
+      int n_features = feature_param_tree["n_features"].get_uint64();
+      int n_levels = feature_param_tree["n_levels"].get_uint64();
+      float scale_factor = feature_param_tree["scale_factor"].get_real();
+
+      feature2d_ = new cv::ORB(n_features, scale_factor, n_levels);
+    }
+#endif
+    else
+    {
+      std::cerr << "Features not implemented for that type: " << feature_type << std::endl;
+      throw;
+    }
+
+  }
+
+  int
+  process(const ecto::tendrils& inputs, const ecto::tendrils& outputs)
+  {
     // Get the DB
     object_recognition_core::db::ObjectDbPtr db =
         object_recognition_core::db::ObjectDbParameters(*json_db_).generateDb();
+
     // Get observations from the DB
     object_recognition_core::db::View view(
         object_recognition_core::db::View::VIEW_OBSERVATION_WHERE_OBJECT_ID);
     view.set_key(*object_id_);
+
     object_recognition_core::db::ViewIterator view_iterator(view, db);
 
     std::vector<cv::Mat> descriptors_all, points_all;
     object_recognition_core::db::ViewIterator iter = view_iterator.begin(),
         end = view_iterator.end();
-    for (; iter != end; ++iter) {
+
+    int counter = 1;
+
+    for(; iter != end; ++iter) {
+
       // Convert the observation to a usable type
       object_recognition_core::prototypes::Observation obs;
       object_recognition_core::db::Document view_element = (*iter);
       obs << &view_element;
 
       // Compute the features/descriptors on the image
-      cv::Mat points, descriptors;
+      cv::Mat /*points,*/ descriptors;
       std::vector<cv::KeyPoint> keypoints;
+
       // TODO actually use the params and do not force ORB
 #if CV_VERSION_MAJOR == 3
-      cv::Ptr<cv::DescriptorExtractor> orb = cv::ORB::create();
-      orb->detectAndCompute(obs.image, obs.mask, keypoints, descriptors);
+      feature2d_->detectAndCompute(obs.image, obs.mask, keypoints, descriptors);
 #else
-      cv::ORB orb;
-      orb(obs.image, obs.mask, keypoints, descriptors);
+      (*feature2d_)(obs.image, obs.mask, keypoints, descriptors);
```

Comment:
Besides, I think it's not coherent that in detection we use ecto_opencv in order to compute the features and then at training time just call the OpenCV C++ API.